lib/HLSL/DxilEliminateOutputDynamicIndexing.cpp

```diff
@@ -0,0 +1,187 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+// Eliminate dynamic indexing on output.                                     //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#include "dxc/HLSL/DxilGenerationPass.h"
+#include "dxc/HLSL/DxilOperations.h"
+#include "dxc/HLSL/DxilSignatureElement.h"
+#include "dxc/HLSL/DxilModule.h"
+#include "dxc/Support/Global.h"
+#include "dxc/HLSL/DxilInstructions.h"
+
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+#include "llvm/IR/IRBuilder.h"
+#include <map>
+
+using namespace llvm;
+using namespace hlsl;
+
+namespace {
+class DxilEliminateOutputDynamicIndexing : public ModulePass {
+private:
+
+public:
+  static char ID; // Pass identification, replacement for typeid
+  explicit DxilEliminateOutputDynamicIndexing() : ModulePass(ID) {}
+
+  const char *getPassName() const override {
+    return "DXIL eliminate ouptut dynamic indexing";
+  }
+
+  bool runOnModule(Module &M) override {
+    DxilModule &DM = M.GetOrCreateDxilModule();
+    bool bUpdated = false;
+    if (DM.GetShaderModel()->IsHS()) {
+      hlsl::OP *hlslOP = DM.GetOP();
+      bUpdated = EliminateDynamicOutput(
+          hlslOP, DXIL::OpCode::StorePatchConstant,
+          DM.GetPatchConstantSignature(), DM.GetPatchConstantFunction());
+    }
+
+    // Skip pass thru entry.
+    if (!DM.GetEntryFunction())
+      return bUpdated;
+
+    hlsl::OP *hlslOP = DM.GetOP();
+
+    bUpdated |=
+        EliminateDynamicOutput(hlslOP, DXIL::OpCode::StoreOutput,
+                               DM.GetOutputSignature(), DM.GetEntryFunction());
+
+    return bUpdated;
+  }
+
+private:
+  bool EliminateDynamicOutput(hlsl::OP *hlslOP, DXIL::OpCode opcode, DxilSignature &outputSig, Function *Entry);
+  void ReplaceDynamicOutput(ArrayRef<Value *> tmpSigElts, Value * sigID, Value *zero, Function *F);
+  void StoreTmpSigToOutput(ArrayRef<Value *> tmpSigElts, unsigned row,
+                           Value *opcode, Value *sigID, Function *StoreOutput,
+                           Function *Entry);
+};
+
+struct CmpSigType {
+  bool operator()(const Value *v0, const Value *v1) const {
+    return cast<ConstantInt>(v0)->getLimitedValue() >
+           cast<ConstantInt>(v1)->getLimitedValue();
+  }
+};
+
+bool DxilEliminateOutputDynamicIndexing::EliminateDynamicOutput(
+    hlsl::OP *hlslOP, DXIL::OpCode opcode, DxilSignature &outputSig,
+    Function *Entry) {
+  ArrayRef<llvm::Function *> storeOutputs =
+      hlslOP->GetOpFuncList(opcode);
+
+  std::map<Value *, Type *, CmpSigType> dynamicSigSet;
```

Comment:
Can use an `llvm::DenseMap<Constant *, Type *>` here since you require the value to be constant and the constants will be uniqued.

Or even just `llvm::DenseMap<unsigned, Type *>` if you don't need the value*.